### PR TITLE
[2.0 -> Master] User Preferences on Sidebar

### PR DESCRIPTION
**Pull request by dillchen**
Allows users to set visibility of in-community sidebar, supersedes #4114 

This PR adds the functionality to respect the user's preference for the sidebar visibility across different screen sizes. If the user has not toggled the sidebar visibility, the app will continue to show or hide the sidebar based on the window size.

With included feature flag which requires .env `FLAG_SIDEBAR_TOGGLE`

Updated video of functionality

https://www.loom.com/share/7089d38fb5144f83986aaa3e04559eec

Here is a video of the functionality

https://www.loom.com/share/d207bcd313f54a26940c7dccef8aab1f

Here's a second video for desktop

https://www.loom.com/share/8d6e94a758874519880a8192048ee3d9?from_recorder=1&focus_title=1

## Test Plan

- Manually tested the sidebar visibility toggle on different screen sizes in both desktop and mobile views.

## Acceptance Criteria
- As a User, I want to be able to see a sidebar open and close on desktop (and content should appear without unexpected visual glitches)
- As a user, while I open and close, the main content should not slide all the way over, rather still remain roughly centered
- As a user, the site should remember if I've opened or closed even if I close the site

## Link to Issue
Closes: #TODO

## Deployment Plan
1. Onto `beta` or `demo` for product acceptance
2. Changes already deployed to `experimental`
3. After deploy onto master, Eng needs to repackage the ToDesktop App for users to use (and distribute the link)

### CODE REVIEW

File: packages/commonwealth/client/scripts/state/index.ts
Comment (masvelio) [inline packages/commonwealth/client/scripts/state/index.ts:196]: for the sake of this PR I guess it is fine, however, keep in mind that we are slowly moving away from this file and this state approach as it is not reactive. Eventually, all this logic from this file should land either in zustand or react-query.

File: packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts
Comment (masvelio) [inline packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts:34]: This is a workaround and we should not use localStorage anymore when using zustand. 
To properly solve this, zustand is providing middleware, in this case, persist middleware, that handles localStorage by default. Note that `devtools` is a middleware, so to make it work, `persist` have to wrap the whole store, similarly to devtools, and as a second argument you need to pass `{name: 'sidebar-store'}` and thats it.

https://docs.pmnd.rs/zustand/recipes/recipes#persist-middleware

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:31]: you have useSidebarStore already in line 24

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:87]: can we use `clsx` for conditional classes? it is way more readable. We use it already in the project

The code would look like this
```
          <div
            className={clsx('body-and-sticky-headers-container', {
              'menu-visible': menuVisible,
              'menu-hidden': !menuVisible,
              resizing,
            })}
          >
```

File: packages/commonwealth/client/scripts/views/components/sidebar/explore_sidebar.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/sidebar/explore_sidebar.tsx:10]: image

File: packages/commonwealth/client/scripts/views/components/sidebar/index.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/sidebar/index.tsx:41]: this is good usecase for clsx

File: packages/commonwealth/client/scripts/views/components/sidebar/index.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/sidebar/index.tsx:23]: image

dont understand why we need `setShouldRender`?

File: packages/commonwealth/client/scripts/views/SublayoutHeader.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/SublayoutHeader.tsx:199]: remove this line

File: packages/commonwealth/client/scripts/views/SublayoutHeader.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/SublayoutHeader.tsx:58]: accessing html nodes like this, and setting classes like this is totally non-react approach. The whole UI logic should be based on state but now we are bypassing state and modify DOM directly, which is anti-pattern.

**Review by masvelio — COMMENTED** [review]
while the new sidebar animation looks great, there are some code specific things that should be fixed. Also I am noticing few bugs in new sidebar that I am trying to list below

**Comment by masvelio** [comment]
There are 2 scrollbars visible and the UX is bad for scrolling on some pages eg on dashboard, you can see that you can scroll the general layout so the navbar disappears. If i want to scroll the feed I need to hover over the feed container to scroll it. That is not desired for sure. 
Similarly on communities page, two different scroll behaviours depending on the mouse current position.

https://github.com/hicommonwealth/commonwealth/assets/14819225/8c48593d-30ee-4158-95bd-1fbc4fe1f78b

**Comment by masvelio** [comment]
mobile seems to completely broken 
- not able to show mobile menu
- content of the sidebar is not visible
- clicking on topic does not close the sidebar and does not show the desired page
- hardly anything is working 

https://github.com/hicommonwealth/commonwealth/assets/14819225/ff386989-6c94-435c-808f-b0caca696c58

**Comment by dillchen** [comment]
> There are 2 scrollbars visible and the UX is bad for scrolling on some pages eg on dashboard, you can see that you can scroll the general layout so the navbar disappears. If i want to scroll the feed I need to hover over the feed container to scroll it. That is not desired for sure. Similarly on communities page, two different scroll behaviours depending on the mouse current position.
> 
>  Kapture.2023-08-22.at.13.47.08.mp4

What browser are you using?

**Comment by dillchen** [comment]
> When I pick community from the sidebar, the sidebar disappears and I am not able to get it back, only reload helps.
> 
>  Kapture.2023-08-22.at.13.42.28.mp4

will fix

**Comment by dillchen** [comment]
> mobile seems to completely broken
> 
> * not able to show mobile menu
> * content of the sidebar is not visible
> * clicking on topic does not close the sidebar and does not show the desired page
> * hardly anything is working
> 
>  Kapture.2023-08-22.at.13.56.48.mp4

These are surprising and looks like I've reintroduced while merging master in.

File: packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts
Comment (dillchen) [reply packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts:34]: Is fixing this specific comment blocking on merge @masvelio ?

**Comment by dillchen** [comment]
https://www.loom.com/share/6d2913b4bba2442e8a6c7990d52f19f2

Update Loom @masvelio

File: packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts
Comment (masvelio) [reply packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts]: I see that you updated it, thanks. It is really important that we follow new standards for state management and we are going away from workarounds and old state pattern

**Comment by masvelio** [comment]
> > There are 2 scrollbars visible and the UX is bad for scrolling on some pages eg on dashboard, you can see that you can scroll the general layout so the navbar disappears. If i want to scroll the feed I need to hover over the feed container to scroll it. That is not desired for sure. Similarly on communities page, two different scroll behaviours depending on the mouse current position.
> > Kapture.2023-08-22.at.13.47.08.mp4
> 
> What browser / OS are you using?, not initially clear how you're getting double scroll bar. This should be addressed separately.
> 
> In the update loom, I've provided I reproduce your behavior. which has always been the case, but very much can be a UX improvement (that is a user is not able to scroll on the feed while they are hovering over the explore column)

I am using mac/chrome but it does not matter here. Double scroll is reproducible on all browsers. I have an external mouse and the scrollbars appear differently when you have mouse hooked up.

I think it should be address it in this PR, because this PR introduces double scrollbar. On master it works well everywhere. Please take a look on [this PR](https://github.com/hicommonwealth/commonwealth/pull/4700/files) because we work on this with Kurtis and I believe that this is a root cause of the problem.

File: packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts
Comment (masvelio) [inline packages/commonwealth/client/scripts/state/ui/sidebar/sidebar.ts:62]: great!

File: packages/commonwealth/client/scripts/views/Sublayout.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/Sublayout.tsx:25]: `'userToggledVisibility' is assigned a value but never used`

**Comment by masvelio** [comment]
on mobile, when I click parent menu item, the menu is hiding, but it shouldn't

https://github.com/hicommonwealth/commonwealth/assets/14819225/bfac0fc5-3daa-4b15-8019-5446c2f0ffe2

**Review by masvelio — APPROVED** [review]
left some small comments. Looks and works much better now. Thanks for applying the feedback but would love to resolve "double scroll issue". Approving in advance.

**Comment by dillchen** [comment]
https://www.loom.com/share/7fd46d81699444738170767f514e6f82

Everything addressed outside of double scroll

**Comment by dillchen** [comment]
Ticketed this separately as a fast follow (we will get this in refinement the next week) => https://github.com/hicommonwealth/commonwealth/issues/4899

Note, I'm unable to reproduce (developing from my laptop atm)

**Comment by masvelio** [comment]
@dillchen As I thought, [this simple CSS change](https://github.com/hicommonwealth/commonwealth/pull/4862/commits/c94084978d797db2014f1d941b21ad1a59a016d6) fixed double scroll issue 💪 Double scroll is not visible anymore neither with trackpad nor external mouse. Mobile also looks good. This means that https://github.com/hicommonwealth/commonwealth/issues/4899 and https://github.com/hicommonwealth/commonwealth/pull/4900 is not relevant anymore. 

From my POV, after making tests pass, this PR is g2g!

**Review by jnaviask — CHANGES_REQUESTED** [review]
Need to get unit/integration tests passing before can approve.

File: packages/commonwealth/client/scripts/state/index.ts
Comment (dillchen) [reply packages/commonwealth/client/scripts/state/index.ts:26]: ToDesktop is a no-code way to release on cross-platform desktop Apps.

Due to a quirk of Desktop Apps on Mac, menu bars are not default draggable.

We import the ToDesktop core package and check for the platform. This allow us to create "adaptive CSS" which makes the menu bar draggable.

File: packages/commonwealth/client/scripts/hooks/useInitApp.ts
Comment (dillchen) [reply packages/commonwealth/client/scripts/hooks/useInitApp.ts:15]: Same as above comment,

Allows us to check if we are using a desktop App, therefore rendering "adaptive" UI

**Comment by jnaviask** [comment]
Marking draft until feedback is addressed.

File: packages/commonwealth/client/scripts/views/menus/create_content_menu.tsx
Comment (dillchen) [reply packages/commonwealth/client/scripts/views/menus/create_content_menu.tsx]: resolved

File: packages/commonwealth/client/scripts/views/components/sidebar/explore_sidebar.tsx
Comment (dillchen) [reply packages/commonwealth/client/scripts/views/components/sidebar/explore_sidebar.tsx]: resolved

File: packages/commonwealth/test/mocha-dev.opts
Comment (dillchen) [reply packages/commonwealth/test/mocha-dev.opts:5]: required to get tests to run, because there is CSS being used in the file for finding breakpoints

**Without**

```
$ NODE_ENV=test ts-mocha --project tsconfig.json './test/unit/**/*.spec.ts'
Applied changes: logLevel=Debug to log groups 'all'.

/Users/dillchen/Coding/hicommonwealth/commonwealth/packages/commonwealth/node_modules/ts-mocha/node_modules/ts-node/src/index.ts:261
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
client/styles/mixins/breakpoints.scss(2,22): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(3,23): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(4,23): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(5,29): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(6,29): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(7,22): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(8,22): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(9,28): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(12,25): error TS1005: ';' expected.
client/styles/mixins/breakpoints.scss(12,52): error TS1351: An identifier or keyword cannot immediately follow a numeric literal.

```
See similar issues on [StackOverflow](https://stackoverflow.com/questions/32236443/mocha-testing-failed-due-to-css-in-webpack)


**With**
You'll see the current file

File: packages/commonwealth/test/util/localStorageUtils.ts
Comment (dillchen) [reply packages/commonwealth/test/util/localStorageUtils.ts:1]: resolving as unnecessary and removed

File: packages/commonwealth/client/scripts/state/index.ts
Comment (dillchen) [reply packages/commonwealth/client/scripts/state/index.ts]: resolved => using a window check now app.isDesktop instead of relying on package function

File: packages/commonwealth/client/scripts/hooks/useInitApp.ts
Comment (dillchen) [reply packages/commonwealth/client/scripts/hooks/useInitApp.ts]: When we are on the mobile native apps or on desktop, there is no reason to check for domains (because custom domains are not allowed)

**Review by jnaviask — APPROVED** [review]
Seems ok code-wise. Some points to note:
- Not sure if "changing page width with collapsed sidebar causes it to uncollapse" is expected behavior or a bug.
- cc @gdjohnson for two items: (1) we should ensure we have all these new feature flags documented, in this case `FLAG_SIDEBAR_TOGGLE`, and (2) we should ensure we have documentation for all the different platforms possible (seems like we have "desktop" and also a variety from capacitor, including "web" and maybe mobile versions?).
- Approving but do not merge, pending Session Keys which should be next release.

**Comment by jnaviask** [comment]
cc @kurtisassad to review E2E test failure

**Comment by jnaviask** [comment]
cc @dillchen to fix merge conflicts then we will merge

**Comment by kurtisassad** [comment]
e2e test looks fine, for whatever reason the test didn't catch the textbox when making a new thread.

Tested it locally and works fine.

**Comment by gdjohnson** [comment]
Just flagging that `feature-flags.ts` doesn't match our code style conventions for file naming.

The /helpers directory in general is a mess, with _three different casing styles_ across only a dozen files: camelCase, snake_case, kebab-case.

We have some documentation now, added by @masvelio, as to best practices for naming (1) folders, (2) components. But we haven't made a call on utils, helpers, and similar ts/tsx files. Might be worth figuring this out + documenting + team-wide announcing, but I'll leave it to the discretion of @jnaviask and @CowMuon

**Pull request merged**
